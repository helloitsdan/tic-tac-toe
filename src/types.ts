export enum BoardPosition {
  TOP_LEFT,
  TOP_CENTRE,
  TOP_RIGHT,
  MIDDLE_LEFT,
  MIDDLE_CENTRE,
  MIDDLE_RIGHT,
  BOTTOM_LEFT,
  BOTTOM_CENTRE,
  BOTTOM_RIGHT,
}

export enum PlayerPiece {
  PLAYER_ONE = 'PLAYER_ONE',
  PLAYER_TWO = 'PLAYER_TWO',
}

export enum MatchName {
  TOP_ROW = 'TOP_ROW',
  MIDDLE_ROW = 'MIDDLE_ROW',
  BOTTOM_ROW = 'BOTTOM_ROW',
  LEFT_COLUMN = 'LEFT_COLUMN',
  CENTRE_COLUMN = 'CENTRE_COLUMN',
  RIGHT_COLUMN = 'RIGHT_COLUMN',
  DIAGONAL_LEFT_TO_RIGHT = 'DIAGONAL_LEFT_TO_RIGHT',
  DIAGONAL_RIGHT_TO_LEFT = 'DIAGONAL_RIGHT_TO_LEFT',
}

export interface PlayerType {
  piece: PlayerPiece
  name: string
}

export type MatchTuple = [MatchName, BoardPosition[]]

export interface PlayerMatchType {
  match: MatchTuple
  piece: PlayerPiece
}

export type GameMovesType = Record<PlayerPiece, BoardPosition[]>

export interface GameStateType {
  players: PlayerType[]
  moves: GameMovesType
  winner: PlayerMatchType
}
