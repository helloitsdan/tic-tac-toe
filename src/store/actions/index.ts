import { PlayerId, BoardPosition, PlayerMatchType } from "../../types"

export const RESET_GAME = 'game/RESET'
export const PLACE_PIECE = 'game/PLACE_PIECE'
export const WIN_GAME = 'game/WIN'
export const CHANGE_PLAYER = 'player/CHANGE'

export interface ResetGameAction {
  type: typeof RESET_GAME
}

export interface PlacePieceAction {
  type: typeof PLACE_PIECE
  payload: {
    player: PlayerId,
    position: BoardPosition
  }
}

export interface WinGameAction {
  type: typeof WIN_GAME,
  payload: PlayerMatchType
}

export interface ChangePlayerAction {
  type: typeof CHANGE_PLAYER,
  payload: PlayerId
}

export type GameActions = ResetGameAction | PlacePieceAction | WinGameAction | ChangePlayerAction

export const resetGame = (): GameActions => ({
  type: RESET_GAME
})

export const placePiece = (player: PlayerId, position: BoardPosition): GameActions => ({
  type: PLACE_PIECE,
  payload: {
    player,
    position
  }
})

export const winGame = (winningMatch: PlayerMatchType): GameActions => ({
  type: WIN_GAME,
  payload: winningMatch,
})

export const changePlayer = (newPlayer: PlayerId): GameActions => ({
  type: CHANGE_PLAYER,
  payload: newPlayer,
})