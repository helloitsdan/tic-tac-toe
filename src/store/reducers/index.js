const board = (defaultState = {}, action) => {
  return defaultState
}

export default board