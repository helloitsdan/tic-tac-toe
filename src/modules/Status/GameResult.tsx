import React, { FunctionComponent, useMemo } from 'react'
import { useSelector } from 'react-redux'
import { getWinner, getPlayers } from '../../store/selectors'

const GameResult: FunctionComponent = () => {
  const winner = useSelector(getWinner)
  const players = useSelector(getPlayers)

  const player = useMemo(() => {
    if (!winner) {
      return null
    }

    return players.all.find(p => p.piece === winner!.player)
  }, [winner, players])

  return player ? (
    <div data-testid="GameResult/Player">
      <span data-testid="GameResult/PlayerName">{player.name}</span> wins!
    </div>
  ) : (
    <div data-testid="GameResult/Draw">A draw...</div>
  )
}

export default GameResult
