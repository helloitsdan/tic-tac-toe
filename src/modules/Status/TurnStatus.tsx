import React, { FunctionComponent, useMemo } from 'react'
import { useSelector } from 'react-redux'
import { getPlayers } from '../../store/selectors'

const TurnStatus: FunctionComponent = () => {
  const { active } = useSelector(getPlayers)
  const players = useSelector(getPlayers)

  const player = useMemo(() => players.all.find(p => p.piece === active), [active, players])

  return (
    <div data-testid="TurnStatus">
      {player ? (
        <>
          <span data-testid="TurnStatus/Player">{player.name}</span>'s turn!
        </>
      ) : (
        <span data-testid="TurnStatus/UnknownPlayer">Your turn!</span>
      )}
    </div>
  )
}

export default TurnStatus
