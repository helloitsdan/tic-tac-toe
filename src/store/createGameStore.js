import { createStore, compose, applyMiddleware } from "redux";
import createSagaMiddleware from "redux-saga";

import reducers from "./reducers";
import gameState from "./sagas/gameState";

const createGameStore = () => {
  const sagaMiddleware = createSagaMiddleware();

  const composeEnhancers =
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
  const store = createStore(reducers, composeEnhancers(applyMiddleware(sagaMiddleware)));

  sagaMiddleware.run(gameState);

  return store;
};

export default createGameStore;
