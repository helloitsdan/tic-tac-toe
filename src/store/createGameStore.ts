import { createStore, compose, applyMiddleware } from 'redux'
import createSagaMiddleware from 'redux-saga'

import reducers from './reducers'
import gameState from './sagas/gameState'

declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__?: typeof compose
  }
}

const createGameStore = () => {
  const sagaMiddleware = createSagaMiddleware()

  const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose
  const store = createStore(reducers, composeEnhancers(applyMiddleware(sagaMiddleware)))

  sagaMiddleware.run(gameState)

  return store
}

export default createGameStore
