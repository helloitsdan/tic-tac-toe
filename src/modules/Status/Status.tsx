import React, { FunctionComponent, useCallback } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { resetGame } from '../../store/actions'
import { getStatus } from '../../store/selectors'
import { GameStatus } from '../../types'
import GameResult from './GameResult'
import TurnStatus from './TurnStatus'

const Status: FunctionComponent = () => {
  const dispatch = useDispatch()
  const onResetGame = useCallback(() => {
    dispatch(resetGame())
  }, [dispatch])

  const gameStatus = useSelector(getStatus)

  return (
    <>
      {gameStatus === GameStatus.FINISHED ? (
        <div data-testid="Status/GameResult" className="u-push-bottom">
          <GameResult />
        </div>
      ) : (
        <div data-testid="Status/TurnStatus" className="u-push-bottom">
          <TurnStatus />
        </div>
      )}

      <nav>
        <button className="c-button" onClick={onResetGame}>
          Start new game!
        </button>
      </nav>
    </>
  )
}

export default Status
