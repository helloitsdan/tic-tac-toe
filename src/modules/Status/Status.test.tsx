import React from 'react'
import { fireEvent, render } from '@testing-library/react'

import Status from './Status'
import { createStore } from 'redux'
import { Provider } from 'react-redux'
import { PlayerId, GameStateType, GameStatus, BoardPosition } from '../../types'

import { resetGame } from '../../store/actions'

jest.mock('../../store/actions')

jest.mock('')

describe('Status', () => {
  const testState: GameStateType = {
    players: {
      active: PlayerId.PLAYER_ONE,
      all: [
        { piece: PlayerId.PLAYER_ONE, name: 'Test Player One' },
        { piece: PlayerId.PLAYER_TWO, name: 'Test Player Two' },
      ]
    },
    status: GameStatus.PLAYING,
    winner: null,
    moves: {
      PLAYER_ONE: [BoardPosition.MIDDLE_CENTRE],
      PLAYER_TWO: [],
    },
  }

  it('renders', () => {
    const { asFragment } = render(
      <Provider store={createStore(() => testState)}>
        <Status />
      </Provider>
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('renders turn status when the game is being played', () => {
    const testStore = createStore(() => ({
      ...testState,
      status: GameStatus.PLAYING,
    }))

    const { getByTestId } = render(
      <Provider store={testStore}>
        <Status />
      </Provider>
    )
    
    expect(getByTestId('Status/TurnStatus')).not.toBeNull()
  })

  it('renders a result when the game has finished', () => {
    const testStore = createStore(() => ({
      ...testState,
      status: GameStatus.FINISHED,
    }))

    const { getByTestId } = render(
      <Provider store={testStore}>
        <Status />
      </Provider>
    )
    
    expect(getByTestId('Status/GameResult')).not.toBeNull()
  })

  it('reset game button dispatches reset action', () => {
    const resetGameMock = resetGame as jest.Mock
    resetGameMock.mockReturnValue({ type: 'PRETEND' })

    const { getByRole } = render(
      <Provider store={createStore(() => testState)}>
        <Status />
      </Provider>
    )
    fireEvent.click(getByRole('button'))
    
    expect(resetGameMock).toHaveBeenCalled()
  })
})
