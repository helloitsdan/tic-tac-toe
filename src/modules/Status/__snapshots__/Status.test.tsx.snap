// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Status renders 1`] = `
<DocumentFragment>
  <div
    class="u-push-bottom"
    data-testid="Status/TurnStatus"
  >
    <div
      data-testid="TurnStatus"
    >
      <span
        data-testid="TurnStatus/Player"
      >
        Test Player One
      </span>
      's turn!
    </div>
  </div>
  <nav>
    <button
      class="c-button"
    >
      Start new game!
    </button>
  </nav>
</DocumentFragment>
`;
